### DISCUSSION TITLE
【腾讯犀牛鸟2026】Penguin-VL ncnn 移植与多平台部署

### DISCUSSION THREAD

**Discussion by czwdysj** [Tencent/ncnn#6821] [Show and tell]
# Penguin-VL ncnn Runtime

项目github仓库链接：[[czwdysj/Penguin-VL-ncnn-Runtime: 一个基于ncnn的推理框架 目前可以推理Penguin-VL这个vlm模型](https://github.com/czwdysj/Penguin-VL-ncnn-Runtime)](https://github.com/czwdysj/Penguin-VL-ncnn-Runtime)

## 任务目标

本项目将 `tencent/Penguin-VL-2B` 从 PyTorch/Hugging Face 推理链路移植到 ncnn，并提供一个可以直接读取图片和任意文本 prompt 的 C++ 多模态推理程序：

正式程序需要完全在 C++ 中完成以下工作：

1. 读取并预处理用户图片。
2. 使用 Qwen tokenizer 编码任意文本 prompt。
3. 构造 Penguin-VL 多模态 chat template。
4. 使用 ncnn vision encoder 将图片转换为 image embeddings。
5. 将 image embeddings 注入文本 token embedding。
6. 使用并行 prefill 建立每一层的 KV cache。
7. 使用 KV cache 逐 token decode，并把 token id 解码成文本。

当前已完成：

- vision encoder、projector 和 Qwen3 LLM 的 ncnn 模型转换。
- C++ 图片预处理、Qwen byte-level BPE tokenizer 和 chat template。
- 512-token parallel prefill 与 `seq_len=1` KV-cache decode。
- 低内存 streamed-weight 推理策略。
- Linux CMake 构建和 CPU 端到端运行。
- ncnn 与 PyTorch greedy decode 的 128-token 输出对齐。

## 目录结构

```text
Penguin-VL-ncnn/
├── assets/
│   ├── ncnn_icon.png
│   │   └── 端到端推理使用的输入图片
│   └── penguinvl_ncnn_pytorch_alignment.png
│       └── ncnn 与 PyTorch 输出对齐截图
├── src/
│   ├── app/
│   │   ├── main.cpp
│   │   │   └── 命令行入口，串联完整多模态推理流程
│   │   └── cli.*
│   │       └── 解析 --image、--prompt、--model 等参数
│   ├── core/
│   │   ├── image_processor.*
│   │   │   └── PNG 解码、resize、normalize、patch flatten 和 vision RoPE
│   │   ├── prompt_builder.*
│   │   │   └── 构造 chat template 并插入 image token 占位符
│   │   └── qwen_tokenizer.*
│   │       └── Qwen byte-level BPE encode/decode
│   └── runtime/
│       ├── ncnn_utils.h
│       │   └── ncnn 网络加载、执行和张量辅助函数
│       ├── vision_encoder.*
│       │   └── 执行 vision embedding、28 层视觉 Transformer、norm 和 projector
│       └── text_decoder.*
│           └── LLM embedding、parallel prefill、KV-cache decode、norm 和 lm_head
├── tools/
│   ├── python/
│   │   ├── run_torch_reference.py
│   │   │   └── 生成 PyTorch 参考输出
│   │   ├── export_pnnx_modules.py
│   │   │   └── 导出 vision encoder 和 projector
│   │   ├── export_penguinvl_llm_modules.py
│   │   │   └── 导出 Qwen3 LLM layer
│   │   ├── export_penguinvl_embed_lmhead_ncnn.py
│   │   │   └── 生成 embedding、norm、lm_head 和 tokenizer 文件
│   │   └── patch_ncnn_kvcache.py
│   │       └── 为 pnnx attention 图添加显式 KV cache 输入输出
│   └── shell/
│       ├── build_ncnn_modules.sh
│       │   └── 转换 vision encoder 和 projector
│       ├── build_llm_kvcache_prefill_modules.sh
│       │   └── 转换 512-token parallel prefill 图
│       ├── build_llm_kvcache_modules.sh
│       │   └── 转换 seq_len=1 KV-cache decode 图
│       └── package_*.sh
│           └── 把转换结果整理成最终模型目录
├── CMakeLists.txt
│   └── Linux/Windows CMake 工程
├── penguinvl
│   └── Linux 快速启动脚本
└── README.md
```

### 测试输入

测试图片为 ncnn 图标：

<img width="256" height="256" alt="f953f5f6a71b150f2ade368f9deaeb01" src="https://github.com/user-attachments/assets/5a712b3b-35ef-4ee3-961b-68b18ca95a83" />

测试 prompt：

```text
描述这个图片
```

运行命令：

```bash
./penguinvl \
  --image assets/ncnn_icon.png \
  --prompt "描述这个图片" \
  --max-new-tokens 128
```

### ncnn 与 PyTorch 对齐结果

<img width="797" height="632" alt="ecf05ec0e3ed38bba58681e32c7255d0" src="https://github.com/user-attachments/assets/c83139e8-49c7-4684-a289-0cbda1608e0c" />

截图上半部分是 C++ ncnn runtime，下半部分是 PyTorch reference。两端使用相同图片、相同 prompt、greedy decode 和最多 128 个新 token，输出文本一致，并最终生成 EOS token `151645`。

ncnn 输出：

```text
这张图片展示了一个室内场景，背景是白色的墙壁，墙上有一些模糊的图案或装饰。
前景中有一只白色的猫咪，它正坐在一个白色的圆桌上。桌子上有一些模糊的物品，
可能是食物或装饰品。猫咪的表情看起来很放松，眼睛半闭着，似乎在享受这个安静的时刻。
整体氛围显得温馨而宁静。
```

### 解码与采样策略

***当前推理采用 greedy argmax 解码，不启用随机采样。每一步经过 `RMSNorm` 和 `lm_head` 得到整个词表的 logits，然后选择概率最高的 token，只有这样才能保证ncnn和pytorch完全一样 但是这样显然效果比较差 模型回答的不对属于模型本身的问题和采样的问题***

## 模型结构

<img width="1021" height="1476" alt="Penguin-VL-2B-framework-cn" src="https://github.com/user-attachments/assets/16417e5f-933b-4aef-81e6-f1c45dc7dac3" />

<img width="1281" height="791" alt="teansformer_block" src="https://github.com/user-attachments/assets/36afa9b6-a7ac-4d77-bde9-0adbe2ff66db" />

### 1. 建立 PyTorch 参考输出

移植前先运行原始 `tencent/Penguin-VL-2B`，保存输入 token、图片张量 shape、image token 数量、生成 token ids 和最终文本：

```bash
python tools/python/run_torch_reference.py \
  --image assets/ncnn_icon.png \
  --prompt "描述这个图片" \
  --max-new-tokens 128
```

后续每完成一个 ncnn 模块，都与这个参考结果比较。排查顺序是输入处理、vision encoder、projector、LLM layer、KV cache、lm_head 和最终 token，避免只根据最终文本猜测错误来源。

### 2. 在 C++ 中处理图片

`image_processor.cpp` 直接读取 PNG，并执行和 PyTorch processor 一致的处理：

```text
PNG RGB 图片
  -> bicubic resize 到 252x252
  -> mean=0.5、std=0.5 归一化
  -> 切分为 18x18 个 14x14 patch
  -> 每个 RGB patch 展平为 3x14x14=588 维
  -> pixel_values shape = [324, 588]
```

程序同时在 C++ 中计算 vision RoPE。图片数据和位置编码一起输入 ncnn vision encoder，不需要 Python 生成中间 `.npy`。

### 3. 把图片转换为 image embeddings

视觉部分被拆成下面几个 ncnn 模块：

```text
[324, 588] pixel_values
  -> vision embedding
  -> 28 个 vision transformer layer
  -> vision norm
  -> projector
  -> [324, 2048] image embeddings
```

这里的 image token 不是 tokenizer 产生的整数 token id，而是 324 个连续的 2048 维向量。projector 的作用是把视觉 hidden size 映射到语言模型 hidden size，使视觉向量可以放入 LLM 的 embedding 序列。

### 4. 在 C++ 中编码任意 prompt

`qwen_tokenizer.cpp` 加载 `vocab.json`、`merges.txt` 和 `tokenizer_config.json`，实现 Qwen byte-level BPE：

```text
用户 UTF-8 文本
  -> special token 识别
  -> byte-level 映射
  -> BPE merge
  -> token ids
```

`prompt_builder.cpp` 再按照 Penguin-VL 的 chat template 组合 system/user/assistant 标记，并在图片位置插入 324 个 `image_token_id=151669`。

LLM embedding 首先把文本 token id 转换为 `[sequence_length, 2048]` 向量，然后把 324 个 image token 占位位置替换为 vision projector 输出。到这里，图片和文字已经融合成同一个 Transformer 输入序列。

### 5. 转换 LLM 为 prefill 和 decode 两类图

语言模型有 28 层。每一层分别导出两种 ncnn 图：

- `seq_len=512` prefill 图：并行处理完整图文 prompt，输出当前层 hidden states 和 K/V cache。
- `seq_len=1` decode 图：只处理最新 token，读取历史 K/V cache，并输出追加后的 cache。

`patch_ncnn_kvcache.py` 会修改 pnnx 生成的 attention 图，增加 `past_key`、`past_value` 输入和 `present_key`、`present_value` 输出。这样 KV cache 的生命周期由 C++ runtime 管理，而不是每次生成 token 都重新运行完整 prompt。

### 6. Parallel prefill

当前测试 prompt 包含 340 个 token。runtime 将 hidden states 右 padding 到 512，然后逐层执行：

```text
512 个 hidden states
  -> layer 0，保存 layer 0 K/V
  -> layer 1，保存 layer 1 K/V
  -> ...
  -> layer 27，保存 layer 27 K/V
  -> 取真实 prompt 最后位置的 hidden state
```

一次处理 512 个位置，比对 340 个 token 分别运行 28 层快很多。prefill 完成后，runtime 会把每层 K/V 从 512 裁剪回真实 prompt 长度 340，删除右侧 padding 对应的 cache。

### 7. 逐 token KV-cache decode

prefill 得到第一个 logits 后，runtime 使用 greedy argmax 选择 token。之后每轮只处理新 token：

```text
最新 token
  -> embedding
  -> layer 0 + layer 0 历史 K/V
  -> layer 1 + layer 1 历史 K/V
  -> ...
  -> layer 27 + layer 27 历史 K/V
  -> RMSNorm
  -> lm_head
  -> argmax 得到下一个 token
```

生成过程达到 `--max-new-tokens` 或遇到 EOS token `151645` 时停止，最后由 Qwen tokenizer 把生成 token ids 解码成 UTF-8 文本。

### 8. 构建和运行

Linux 编译：

```bash
cmake -S . -B build/linux-native -DNCNN_ROOT=/path/to/ncnn
cmake --build build/linux-native -j2
```

Windows PowerShell 编译：

```powershell
cmake -S . -B build/windows-native -DNCNN_ROOT=D:\path\to\ncnn
cmake --build build/windows-native --config Release -j 2
```

CMake 支持 ncnn release 根目录和 Windows `x64/`、`x86/`、`arm64/` 子目录布局。MSVC 构建启用 UTF-8，并定义 `NOMINMAX`。

### 1. pnnx 无法直接稳定转换完整多模态模型

原始模型包含 Hugging Face remote code、动态 shape、FlashAttention/SDPA、动态 vision RoPE 和 KV cache。直接转换完整模型会产生 pnnx 无法稳定解析或 ncnn 无法正确执行的图。

解决办法是按计算边界拆分模型：vision embedding、vision layer、vision norm、projector、LLM layer、text norm 和 lm_head 分别转换。vision patch convolution 改写成数学等价的 Linear，attention 改写为 pnnx 更容易识别的算子组合，RoPE、mask 和 cache 由 C++ runtime 显式提供。

### 2. Tokenizer 必须逐 token 对齐

只要空格、换行、UTF-8 byte 映射、BPE merge 或 special token 中有一个细节和 Hugging Face 不一致，prompt token 序列就会变化，后面所有输出都无法对齐。

解决办法是完整实现 Qwen byte-level BPE，并把调试顺序固定为：先对比 input ids，再对比 image token 位置，最后才检查模型输出。不能只看解码后的中文是否相似。

### 3. 图片 token 不是普通文本 token

文本 token 先通过 embedding table 查表；图片位置虽然在 input ids 中表现为 `151669`，真正输入 LLM 的内容却必须是 vision projector 生成的连续向量。若只把 `151669` 当普通词表 id 查 embedding，模型就看不到图片。

解决办法是在文本 embedding 完成后，根据 image token positions 精确替换 324 行 embedding，并检查 image token 数量必须与 projector 输出行数一致。

### 4. 512 prefill 的 padding 导致 PyTorch 与 ncnn 后续 token 不一致

真实 prompt 长度为 340，prefill 图长度为 512。如果直接保留全部 K/V，decode 会把 172 个 padding 位置当成真实历史上下文，attention 范围和 RoPE position 都会错位。其表现通常是第一个 token 正确，后续 token 逐渐偏离。

解决办法是在 prefill 后，把全部 28 层的 key cache 和 value cache 从 512 裁剪回 340，再从 position 340 开始 decode。修复后 ncnn 与 PyTorch 的 128-token greedy 输出一致。

### 5. 内存不足以常驻全部 LLM 权重

28 层 fp16 ncnn 权重同时加载会超过当前约 8GB 的 WSL 内存并触发 OOM。完全逐 token 重跑 prompt 虽然省内存，但速度无法接受。

最终采用 streamed-weight KV runtime：prompt 使用 512 parallel prefill；每次只加载当前 layer，执行后立即释放；decode 复用 KV cache，但仍按层流式加载权重。这样峰值内存主要由当前层权重、hidden states 和 KV cache 构成。

代价是每生成一个 token 都需要从磁盘读取 28 层权重，因此当前实现偏慢。内存充足的平台可以增加 resident-weight 模式，把所有 layer 常驻内存来提高 decode 速度。
